Return first hit, simplify code

sudo-update opened the pull request:
## PR description
"BlockchainIndex" returned 10 hits, but only the first one was observed and returned. I simplified the code to returned first hit only and removed the for loops.
## Fixed Issue(s)
fixes #267

atoulme commented on eth-repository/src/main/kotlin/org/apache/tuweni/eth/repository/BlockchainIndex.kt:
```diff
@@ -483,10 +483,8 @@ class BlockchainIndex(private val indexWriter: IndexWriter) : BlockchainIndexWri
-        docs += doc
-      }
+      val doc = searcher.doc(topDocs.scoreDocs.elementAt(0).doc, setOf("_id") + fields)
+      docs += doc
```
Maybe just return the doc instead of a list here.

sudo-update replied:
It will change the return type from List<Document> to a single Document, is that okay?

atoulme replied:
Looks ok from here.

sudo-update replied:
When I make the above change, I get this error:
Task :eth-repository:compileKotlin FAILED
e: /home/gradle/eth-repository/src/main/kotlin/org/apache/tuweni/eth/repository/BlockchainIndex.kt: (485, 14): Type mismatch: inferred type is Document! but List<Document> was expected

FAILURE: Build failed with an exception.

atoulme commented:
Can you also take care of line 554? 10 is inlined there, should be HITS. And there's a for loop right below.

sudo-update commented:
> Can you also take care of line 554? 10 is inlined there, should be HITS. And there's a for loop right below.

I have fixed the issue on line 554 as well.

However I cannot change the return type to a doc from a list (as requested) due to a build exception.

atoulme reviewed (approved):
lgtm

sudo-update commented:
> lgtm

is there anything else I should do, in order to get this pull request merged?

atoulme commented:
You have to get a successful build. Not sure why it fails right now.

sudo-update commented:
I am unsure why, but GitHub was saying I was behind main, so I merged from main, just to make sure I was working from the latest codebase.  I have changed EthHandlerTest.kt to not loop through 10 indices. On my end, I am building successfully.

sudo-update commented:
I am unable to start the workflow check, as a first time contributor

atoulme commented:
Running the ci now.

atoulme commented on devp2p-eth/src/test/kotlin/org/apache/tuweni/devp2p/eth/EthHandlerTest.kt:
```diff
@@ -128,22 +128,22 @@ class EthHandlerTest {
+          Bytes32.random(),
+          32L, LogsBloomFilter(), emptyList()
+        ),
+        txIndex, newBlock.body.transactions.hash, newBlock.header.hash
```
this is not going to compile, the list of transactions doesn't have a hash, each transaction has its own hash.
Is there a reason for modifying this test? Seems unrelated to your changes.

atoulme commented on eth-repository/src/main/kotlin/org/apache/tuweni/eth/repository/BlockchainIndex.kt:
```diff
@@ -553,16 +550,16 @@ class BlockchainIndex(private val indexWriter: IndexWriter) : BlockchainIndexWri
-      for (hit in topDocs.scoreDocs) {
-        val doc = searcher.doc(hit.doc, setOf("_id"))
-        val bytes = doc.getBinaryValue("_id")
+      val doc = searcher.doc(topDocs.scoreDocs.elementAt(0).doc, setOf("_id"))
```
What if scoreDocs is empty? Would be good to check the element exists, and return null early otherwise.

sudo-update replied:
### Looking into it now

sudo-update commented:
I think I have fixed the issue now, I am getting build successful on my end.  I have requested re-review, hopefully I have succeeded!

I reverted my changes to EthHandlerTest, you were correct.  I did not need to change anything there.  I also made the changes you suggested, to return null when scoreDocs was empty.  That solved the build failure, I think that was the issue.

sudo-update commented:
Awesome! It passed all checks! Does it look good to you?

atoulme reviewed (approved):
LGTM
